core/src/main/java/com/graphhopper/routing/util/parsers/OSMTurnRelationParser.java

```diff
@@ -0,0 +1,194 @@
+        if (!turnRelation.isVehicleTypeConcernedByTurnRestriction(restrictions))
+            return;
+
+        addRelationToTCStorage(turnRelation, turnCostFlags, map, graph);
+    }
+
+    private EdgeExplorer getInExplorer(Graph graph) {
+        return cachedInExplorer == null ? cachedInExplorer = graph.createEdgeExplorer(DefaultEdgeFilter.inEdges(accessEnc)) : cachedInExplorer;
+    }
+
+    EdgeExplorer getOutExplorer(Graph graph) {
+        return cachedOutExplorer == null ? cachedOutExplorer = graph.createEdgeExplorer(DefaultEdgeFilter.outEdges(accessEnc)) : cachedOutExplorer;
+    }
+
+    /**
+     * Add the specified relation to the TurnCostStorage
+     *
+     * @return a collection of turn cost entries which can be used for testing
+     */
+    Collection<TCEntry> addRelationToTCStorage(OSMTurnRelation osmTurnRelation, IntsRef turnCostFlags,
+                                               ExternalInternalMap map, Graph graph) {
+        TurnCostStorage tcs = graph.getTurnCostStorage();
+        int viaNode = map.getInternalNodeIdOfOsmNode(osmTurnRelation.getViaOsmNodeId());
+        EdgeExplorer edgeOutExplorer = getOutExplorer(graph), edgeInExplorer = getInExplorer(graph);
+
+        try {
+            int edgeIdFrom = EdgeIterator.NO_EDGE;
+
+            // get all incoming edges and receive the edge which is defined by fromOsm
+            EdgeIterator iter = edgeInExplorer.setBaseNode(viaNode);
+
+            while (iter.next()) {
+                if (map.getOsmIdOfInternalEdge(iter.getEdge()) == osmTurnRelation.getOsmIdFrom()) {
+                    edgeIdFrom = iter.getEdge();
+                    break;
+                }
+            }
+
+            if (!EdgeIterator.Edge.isValid(edgeIdFrom))
+                return Collections.emptyList();
+
+            final Collection<TCEntry> entries = new ArrayList<>();
+            // get all outgoing edges of the via node
+            iter = edgeOutExplorer.setBaseNode(viaNode);
+            // for TYPE_ONLY_* we add ALL restrictions (from, via, * ) EXCEPT the given turn
+            // for TYPE_NOT_*  we add ONE restriction  (from, via, to)
+            while (iter.next()) {
+                int edgeId = iter.getEdge();
+                long wayId = map.getOsmIdOfInternalEdge(edgeId);
+                if (edgeId != edgeIdFrom && osmTurnRelation.getRestriction() == OSMTurnRelation.Type.ONLY && wayId != osmTurnRelation.getOsmIdTo()
+                        || osmTurnRelation.getRestriction() == OSMTurnRelation.Type.NOT && wayId == osmTurnRelation.getOsmIdTo() && wayId >= 0) {
+                    final TCEntry entry = new TCEntry(edgeIdFrom, viaNode, iter.getEdge());
+                    IntsRef tcFlags = new IntsRef(turnCostFlags.length);
+                    getTurnCostEnc().setDecimal(false, tcFlags, Double.POSITIVE_INFINITY);
+                    entries.add(entry);
+                    tcs.setTurnCost(tcFlags, edgeIdFrom, viaNode, iter.getEdge());
```

Comment:
Why can we not use tcs.set(...) here and why do we (need to) create a new IntsRef everytime we are here ?

Reply:
Ah, yes. It is indeed no longer required to create a new IntsRef now that we do no collect the TurnCostEntry. And will also try to use tcs.set...